app/plugin_requests.go

```diff
@@ -40,6 +41,42 @@ func (a *App) ServePluginRequest(w http.ResponseWriter, r *http.Request) {
 	a.servePluginRequest(w, r, hooks.ServeHTTP)
 }
 
+// ServePluginStaticRequest serves static plugin files
+// at the URL http(s)://$SITE_URL/plugins/$PLUGIN_ID/public/{anything}
+func (a *App) ServePluginStaticRequest(w http.ResponseWriter, r *http.Request) {
+	// Static files rarely change and may be cached indefinitely. Though,
+	// it is recommended that it's not more than one year...
+	// RFC 7234 5.3 https://tools.ietf.org/html/rfc7234#section-5.3
+	w.Header().Set("Cache-Control", "max-age=31556926, public")
```

Comment:
Keep in mind that with this kind of cache directive, it will be effectively impossible for a plugin to "fix" existing assets even when a new version of the plugin is installed. To address that, I think it's imperative to either:
1) embed the plugin version into the URL (effectively busting the cache)
2) document a strategy or provide tooling that forces a name change to assets when building the plugin

Reply:
I'm game for either of those. I started out thinking we could just document the one-year cache directive and the plug author could manage their own cache busting with querystring params or file renames. But I like the idea of providing our own mechanism to keep authors out of trouble.

Reply:
Thanks for bringing this point up @lieut-data, I've not thought about this.

How would 1. exactly work? Would we redirect `http://$SITE_URL/plugin/$PLUGIN_ID/public/` to `http://$SITE_URL/plugin/$PLUGIN_ID/$PLUGIN_VERSION/public/`?

Reply:
For Option 1, I had been imagining some plugin-helper to build a versioned path to the assets (e.g. `getPublicURL(asset string)` and an equivalent for javascript) that would automatically include the version, but your idea of automatically redirecting might work as well. Might want to check browser caching behaviour across a redirect like that.

Reply:
It looks like I started some trouble introducing this cache header. Something to keep in mind, we can change URLs on the backend all we want, but the browser won't know to request the new URL. We have a few options:

1. Don't cache. (My least favorite option)
2. Use eTags. I am a new contributor, so I apologize if you guys are already aware of this and Mattermost already has it - but this is a built-in feature of browsers where we send an etag header with each request that is the same for each version of the file. The browser will then send a request to see if the etag has changed. When we receive that request we either send a 304 Not Modified, or we send the new file content with a new etag. This requires us to "know" when the file content changes and keep record of an etag. You can forego the "keeping track of the etag" part by using a date modified timestamp of the file as the etag. (Definitely more robust, but it's at the expense of making a request)
3. We let whatever is rendering the front-end handle its own cache busting. Usually, this is done with a cachebuster query param. (This is actually my favorite. It's the least amount of requests to the server and it allows the author to invent their own cachebusting that suits them. We can document this in best plugin practices and even have a working sample in the demo that theyy can copy) 
```
http://<mattermostserver>/plugins/com.mattermost.sample/public/hello.html?cachebuster=<cachebusterkey>

or just...

http://<mattermostserver>/plugins/com.mattermost.sample/public/hello.html?<cachebusterkey>
```

Reply:
Hey @happygaijin,

Thank you very much for the detailed comment!

I'm not sure I understand 3. correct. Could you elaborate "We let whatever is rendering the front-end handle its own cache busting"? Are you referring to the webapp?

"It allows the author to invent their own cachebusting that suits them." How would the plugin dev to this? How would he/she define a cachebuster? Changing the filename?

Reply:
**"We let whatever is rendering the front-end handle its own cache busting"**
I was under the impression that this public folder issue was put forth so plugins could reference their own public files. So I am referring to whatever is doing the "referencing" of these files in the first place - the plugin HTML/JavaScript I assume. 

**"It allows the author to invent their own cachebusting that suits them."**
I gave a few URL examples above. You could either do this by querystring or by versioned folders. There are other cachebusting techniques, but they are generally used when the server is the one generating the URLs on render. Which we probably wouldn't be, right? These files would likely be referenced in some HTML or JavaScript written by the plugin author. So that rules out mattermost rendering the URLs. Unless we do some sort of regular expression parsing of the plugin rendering output and replace URLs with cachebusted versions.

The initial point that was made was "[caching] it will be effectively impossible for a plugin to "fix" existing assets even when a new version of the plugin is installed".  Before installing a new version of the plugin, the plugin author can manually update the URLs in the HTML/JavaScript (i.e. change the {cachebustervalue} in the above examples. If s/he decides to use querystrings, that is all that will be required. If they decide to use a folder, then they'd need to put the updated file in a new subfolder beneath /public in addition to updating any referencing URLs.

If there are too many references to manually update the URLs in the HTML/JavaScript, then there can be a client-side JavaScript-based routine for including these files. If they are link, img, or script src attributes, they can be run through a Javascript function that "includes" those files on the page with the latest cachebustervalue. <a> tag hrefs can be handled onclick. I suggest we include an example of this in the plugin sample and demo.

Reply:
Maybe it wasn't clear already, but if you change a querystring parameter value, it is considered a different request by the browser and the browser will hit the server. So simply updating ?v=1 to ?v=2 will prompt the browser to re-request the file.

Reply:
I'd actually propose going with the no-cache option for now for simplicity: these are assets likely to be loaded but once per session (since we're a SPA), and plugin authors can build their own caching strategies if its important. The browser also has its own heuristics that avoid needlessly loading assets when there is no explicit caching directive.

Reply:
👍